hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsOutputStream.java

```diff
@@ -168,25 +199,153 @@ public AbfsOutputStream(AbfsOutputStreamContext abfsOutputStreamContext)
     this.lease = abfsOutputStreamContext.getLease();
     this.leaseId = abfsOutputStreamContext.getLeaseId();
     this.executorService =
-        MoreExecutors.listeningDecorator(abfsOutputStreamContext.getExecutorService());
+        MoreExecutors.listeningDecorator(
+            abfsOutputStreamContext.getExecutorService());
     this.cachedSasToken = new CachedSASToken(
         abfsOutputStreamContext.getSasTokenRenewPeriodForStreamsInSeconds());
     this.outputStreamId = createOutputStreamId();
-    this.tracingContext = new TracingContext(abfsOutputStreamContext.getTracingContext());
+    this.tracingContext = new TracingContext(
+        abfsOutputStreamContext.getTracingContext());
     this.tracingContext.setStreamID(outputStreamId);
     this.tracingContext.setOperation(FSOperationType.WRITE);
     this.ioStatistics = outputStreamStatistics.getIOStatistics();
     this.blockFactory = abfsOutputStreamContext.getBlockFactory();
-    this.blockSize = bufferSize;
+    this.isDFSToBlobFallbackEnabled
+        = abfsOutputStreamContext.isDFSToBlobFallbackEnabled();
     // create that first block. This guarantees that an open + close sequence
     // writes a 0-byte entry.
-    createBlockIfNeeded();
+    this.serviceTypeAtInit = this.currentExecutingServiceType =
+        abfsOutputStreamContext.getIngressServiceType();
+    this.clientHandler = abfsOutputStreamContext.getClientHandler();
+    createIngressHandler(serviceTypeAtInit,
+        abfsOutputStreamContext.getBlockFactory(), bufferSize);
+    createBlockIfNeeded(position);
   }
 
+  /**
+   * Retrieves the current ingress handler.
+   *
+   * @return the current {@link AzureIngressHandler}.
+   */
+  public AzureIngressHandler getIngressHandler() {
+    return ingressHandler;
+  }
+
+  /**
+   * Creates an ingress handler based on the provided service type and other parameters.
+   *
+   * @param serviceType the type of service, either BLOB or DFS.
+   * @param blockFactory the factory to create data blocks.
+   * @param bufferSize the buffer size for the ingress handler.
+   * @return the created {@link AzureIngressHandler}.
+   * @throws IOException if there is an error creating the ingress handler.
+   */
+  private AzureIngressHandler createIngressHandler(AbfsServiceType serviceType,
+      DataBlocks.BlockFactory blockFactory,
+      int bufferSize) throws IOException {
+    this.client = clientHandler.getClient(serviceType);
+    if (serviceType == AbfsServiceType.BLOB) {
+      ingressHandler = new AzureBlobIngressHandler(this, blockFactory,
+          bufferSize, eTag, clientHandler);
+    } else if (isDFSToBlobFallbackEnabled) {
+      if (client.getAbfsConfiguration().isSmallWriteOptimizationEnabled()) {
```

Comment:
Just so that config logic source remains same, can below be done.
Store -> AbfsOutputStreamContext :
isSmallWriteOptimizationEnabled();
getIngressServiceType();
doesDefaultServiceTypeSupportSmallWriteOptimization();

Reply:
makes sense, added